File: app/src/main/java/catchla/yep/activity/ProviderOAuthActivity.java
package catchla.yep.activity;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.net.http.SslError;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.webkit.SslErrorHandler;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import java.util.HashMap;
import java.util.Map;

import catchla.yep.BuildConfig;
import catchla.yep.Constants;
import catchla.yep.R;
import catchla.yep.model.TokenAuthorization;
import catchla.yep.util.Utils;
import catchla.yep.util.YepAPIFactory;

/**
 * Created by mariotaku on 15/6/3.
 */
public class ProviderOAuthActivity extends ContentActivity implements Constants {
    private WebView mWebView;

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_provider_oauth);
        final Intent intent = getIntent();
        final String providerName = intent.getStringExtra(EXTRA_PROVIDER_NAME);
        final String url = YepAPIFactory.getProviderOAuthUrl(providerName);
        final Map<String, String> headers = new HashMap<>();
        headers.put("Authorization", TokenAuthorization.getAuthorizationHeaderValue(
                YepAPIFactory.getAuthToken(this, Utils.getCurrentAccount(this))));
        setTitle(getString(R.string.sign_in_to_provider_name, Utils.getProviderName(this, providerName)));
        mWebView.loadUrl(url, headers);
        mWebView.setWebViewClient(new WebViewClient() {
            @Override
            public void onReceivedError(final WebView view, final int errorCode, final String description, final String failingUrl) {
                super.onReceivedError(view, errorCode, description, failingUrl);
            }

            @Override
            public boolean shouldOverrideUrlLoading(final WebView view, final String url) {
                final Uri uri = Uri.parse(url);
                if (YepAPIFactory.isAPIUrl(uri)) {
                    if (YepAPIFactory.isAuthSuccessUrl(url)) {
                        finish();
                    } else if (YepAPIFactory.isAuthFailureUrl(url)) {
                        // TODO: Show error message
                    } else {
                        view.loadUrl(url, headers);
                    }
                    return true;
                }
                System.identityHashCode(url);
                return super.shouldOverrideUrlLoading(view, url);
            }

            @Override
            public void onPageFinished(final WebView view, final String url) {
                super.onPageFinished(view, url);
            }

            @Override
            public void onPageStarted(final WebView view, final String url, final Bitmap favicon) {
                super.onPageStarted(view, url, favicon);
            }

            @Override
            public void onReceivedSslError(final WebView view, @NonNull final SslErrorHandler handler, final SslError error) {
                if (BuildConfig.DEBUG) {
                    handler.proceed();
                } else {
                    handler.cancel();
                }
            }
        });
        final WebSettings settings = mWebView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setBlockNetworkLoads(false);
        settings.setBlockNetworkImage(false);
    }

    @Override
    public void onContentChanged() {
        super.onContentChanged();
        mWebView = (WebView) findViewById(R.id.webview);
    }

    @Override
    protected void onPause() {
        mWebView.onPause();
        super.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mWebView.onResume();
    }

    @Override
    protected void onDestroy() {
        mWebView.destroy();
        super.onDestroy();
    }
}

File: src/main/java/catchla/yep/activity/ProviderOAuthActivity.kt
package catchla.yep.activity

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.net.Uri
import android.net.http.SslError
import android.os.Bundle
import android.support.v4.view.MenuItemCompat
import android.view.Menu
import android.view.View
import android.webkit.SslErrorHandler
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Toast
import catchla.yep.BuildConfig
import catchla.yep.Constants
import catchla.yep.R
import catchla.yep.model.Provider
import catchla.yep.util.YepAPIFactory
import java.util.*

/**
 * Created by mariotaku on 15/6/3.
 */
class ProviderOAuthActivity : ContentActivity(), Constants {
    private lateinit var mWebView: WebView
    private lateinit var mLoadProgress: View

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_provider_oauth)
        val intent = intent
        val providerName = intent.getStringExtra(Constants.EXTRA_PROVIDER_NAME)
        val url = YepAPIFactory.getProviderOAuthUrl(providerName)
        val headers = HashMap<String, String>()
        headers.put("Authorization", YepAPIFactory.getAuthorizationHeaderValue(
                YepAPIFactory.getAuthToken(this, account)))
        title = getString(R.string.sign_in_to_provider_name, Provider.getProviderName(this, providerName))
        mWebView.loadUrl(url, headers)
        mWebView.setWebViewClient(object : WebViewClient() {

            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                val uri = Uri.parse(url)
                if (YepAPIFactory.isAuthSuccessUrl(url)) {
                    finish()
                } else if (YepAPIFactory.isAuthFailureUrl(url)) {
                    Toast.makeText(this@ProviderOAuthActivity, R.string.unable_to_connect, Toast.LENGTH_SHORT).show()
                    finish()
                } else if (YepAPIFactory.isAPIUrl(uri)) {
                    view.loadUrl(url, headers)
                    return true
                }
                return super.shouldOverrideUrlLoading(view, url)
            }

            override fun onPageFinished(view: WebView, url: String) {
                super.onPageFinished(view, url)
                mLoadProgress.visibility = View.GONE
            }

            override fun onPageStarted(view: WebView, url: String, favicon: Bitmap) {
                super.onPageStarted(view, url, favicon)
                mLoadProgress.visibility = View.VISIBLE
            }

            override fun onReceivedSslError(view: WebView, handler: SslErrorHandler, error: SslError) {
                if (BuildConfig.DEBUG) {
                    handler.proceed()
                } else {
                    handler.cancel()
                }
            }
        })
        val settings = mWebView.settings
        settings.javaScriptEnabled = true
        settings.blockNetworkLoads = false
        settings.blockNetworkImage = false
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_provider_oauth, menu)
        mLoadProgress = MenuItemCompat.getActionView(menu.findItem(R.id.load_progress))
        return true
    }

    override fun onContentChanged() {
        super.onContentChanged()
        mWebView = findViewById(R.id.webview) as WebView
    }

    override fun onPause() {
        mWebView.onPause()
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        mWebView.onResume()
    }

    override fun onDestroy() {
        mWebView.destroy()
        super.onDestroy()
    }
}
